### DISCUSSION TITLE
[New Primitive] `FileUpload`

### DISCUSSION THREAD

**Discussion by colmtuite** [radix-ui/primitives#3305] [Ideas]
### Overview

The native `<input type="file">` is not customisable. To get around this, I often see people doing:

```
<label>
  <button>
  <input type="file" style="hide this element visually">
</label>
```

However, this approach doesn't render the filename, which is bad for a11y. Rendering the filename requires JS

### Who does this impact? Who is this for?

Anyone who wants custom UI and uses forms, which is quite a lot of people.

### Additional context

We could also provide a part to render the filesize (which I think is available through the browser API).

### Examples
https://baseweb.design/components/file-uploader

**Comment by asherccohen** [comment]
A good example of use cases can be found on Uber's base web component lib

**Comment by hjonasson** [comment]
Is there any movement on this?

**Comment by benoitgrelard** [comment]
> Is there any movement on this?

No we haven't started any work on this specific primitive.

**Comment by musdotdigital** [comment]
would be a great addition.

**Comment by vermeerenmaxime** [comment]
Can't wait for this (:

**Comment by Dannymx** [comment]
This is definitely needed.

**Comment by pipifix** [comment]
This would be a great addition. Can't wait for this.

**Comment by ederbiason** [comment]
Really useful component, would be a greate addition fr. letss go

**Comment by dexhorthy** [comment]
+1 🙏

**Comment by demarchenac** [comment]
+2 🙏🏻

**Comment by rinorzk** [comment]
+1 🙏🏼

**Comment by samantha-wong** [comment]
+1 🙏

**Comment by trocoli96** [comment]
+1

**Comment by bittercoder** [comment]
This would be super helpful 👍 +1

**Comment by spencer-rafada** [comment]
This would be super helpful +1

**Comment by ereburg** [comment]
+1

**Comment by MaxRojas1990** [comment]
+1

**Comment by rbrahier17** [comment]
+1

**Comment by jlarmstrongiv** [comment]
Another example is https://ark-ui.com/docs/components/file-upload based on https://zagjs.com/components/react/file-upload

**Comment by xdev00** [comment]
is there any update on this ?

**Comment by michaelparkadze** [comment]
Would be a great addition! +0.30000000000000004

**Comment by TareqRafed** [comment]
yes please

**Comment by mondial7** [comment]
+1

**Comment by rsxdalv** [comment]
I will probably use React Aria because this project does not support file input.

**Reply by rsxdalv** [reply]
Now that this issue has been converted into a discussion, my trust has fallen, I'm assuming this will not be implemented.

**Comment by 3umaGH** [comment]
Please?

**Comment by Eigilak** [comment]
Saaame

**Comment by suitux** [comment]
Please!

**Comment by sheikhaaliya** [comment]
Can any body let me know about the file upload thing in radix UI.

**Comment by lsalehar** [comment]
+1

**Comment by mdecrocemovson** [comment]
plz!

**Comment by tdeckardt** [comment]
+1

**Comment by adrian-burlacu-SA** [comment]
bumping in 2025 for [Reka UI](https://reka-ui.com/)

**Comment by AuroraTea** [comment]
2026...
